---
title: "Class 14: Transcriptomics and the Analysis of RNA- Seq data"
author: "Vicky Vu"
date: "2/20/2020"
output: github_document
---

##Import countData and colData from the class website.
The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. The anti-inflammatory effects on airway smooth muscle (ASM) cells has been known for some time but the underlying molecular mechanisms are unclear.

```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```
Now take a look at each.
```{r}
head(counts)
```

```{r}
head(metadata)
```


How many genes are we dealing with?
```{r}
nrow(counts)
```
We want to calculate the average counts per gene for control samples and for the drug treated samples. We will then compare these calies to see if there is a differnce overall between control and treted samples for a given gene. 


This is our metadata info that tells me where our control and treated samples are. (i.e which colums in 'counts')

## Toy Differential gene expression 

To access the control only rows in the metadata 
```{r}
#metadata$dex == "control"

#this line of code gives of the id of the control
control <- metadata[metadata$dex == "control",]
#metadata$id[metadata$dex == "control"] is also acceptable
control
```

Now use the $id of the control data to access the control colums in 'counts' and calculat the average. 
```{r}
control.mean <- rowSums( counts[ ,control$id] ) / nrow(control)
names(control.mean) <- counts$ensgene
```

Now let's do the same for the "treated" samples. 
```{r}
treated <- metadata[metadata$dex == "treated",]
treated

treated.mean <- rowSums( counts[,treated$id] ) / nrow(treated)

```



we will combine out meancount data for bookkeeping purposes.
```{r}
meancounts <- data.frame(control.mean, treated.mean)

```

Plotting 

```{r}
plot (meancounts, log = "xy" )
```

We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])

head(meancounts)

```

Remove any gene that has zero average counts. 
```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

how many genes do I have left?

```{r}
nrow(mycounts)
```


A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)

```
```{r}
sum(up.ind)
sum ( down.ind)
```

## Adding annotation data 



## Do this properly with DESeq2 Analysis 
DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. Bioconductor is a project to provide tools for analyzing high-throughput genomic data including RNA-seq, ChIP-seq and arrays.


```{r}
library( DESeq2)
```
```{r}
citation("DESeq2")
```

Setup input of for DESeq for specifying countadata, metadata, and the design ( i.e. which comparasions we are interested in)

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

DESeq pipeline 

Next, let’s run the DESeq pipeline on the dataset, and reassign the resulting object back to the same variable.

```{r}
dds <- DESeq (dds)
```
```{r}
res <- results (dds)

```

```{r}
summary (res)

resOrdered<- res[order(res$pvalue),] #this orders theresult by the smallest p-value 
```
The results function contains a number of arguments to customize the results table. By default the argument alpha is set to 0.1. If the adjusted p value cutoff will be a value other than 0.1, alpha should be set to that value:
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```
```{r}
resSig05 <- subset(as.data.frame(res), padj < 0.05)
nrow(resSig05)
head(resSig05)
```

```{r}
ord <- order ( resSig05$padj)
head ( resSig05 [ord,])
```


## Data Visualization 
Volcano Plots 






```{r}
mycols <- rep("gray", nrow(res))
mycols[res$log2FoldChange > 2] <- "blue"
mycols[res$log2FoldChange < -2] <- "blue"


plot( res$log2FoldChange,  -log(res$padj), col = mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)")

abline ( v= c(2,-2), lty=2)
abline (h=-log(0.05))
```

